### ISSUE TITLE
Lab 6 level2 keyablemap 

### ISSUE THREAD

**Issue by zeentan22** [nus-cs2030/2122-s2#105]
# Summary

Description of the question / problem here.
I have been trying to compile the code as below however the error java.lang.String cannot be converted to java.lang.ClassLoader keeps appearing

My code is as of below

This is my KeyableMap constructor
![github2](https://user-images.githubusercontent.com/98390645/159027623-2e89d406-abf4-4d33-a2cb-2a158b60f809.png)

And this is my Module constructor which extends the KeyableMap class
![github1](https://user-images.githubusercontent.com/98390645/159027150-90317fc9-2538-4502-8de4-7a499918b4f5.png)

However when I run jshell, this error appears.
![github4](https://user-images.githubusercontent.com/98390645/159027660-c83adf9b-d392-4190-bccc-0834d33a161a.png)

**Comment by eesung00** [comment]
Perhaps your Module need to implements keyable too. I did it and it works for module. But I have some issue with the Student class. Anyone knows why?
![image](https://user-images.githubusercontent.com/90175879/159032597-bb81e642-ee35-4e40-90a3-042bb921661a.png)
![image](https://user-images.githubusercontent.com/90175879/159032699-7e69832c-6dda-4d18-b764-98aa3666338c.png)

**Comment by SpideyMumu** [comment]
It seems like I got the exact same error and posted about it here a few minutes after you haha. Hopefully someone can help us with it soon!

**Comment by LauWyeRock** [comment]
> Perhaps your Module need to implements keyable too. I did it and it works for module. But I have some issue with the Student class. Anyone knows why? ![image](https://user-images.githubusercontent.com/90175879/159032597-bb81e642-ee35-4e40-90a3-042bb921661a.png) ![image](https://user-images.githubusercontent.com/90175879/159032699-7e69832c-6dda-4d18-b764-98aa3666338c.png)

I think your error is because your Module should not be implementing Keyable

**Comment by LauWyeRock** [comment]
@zeentan22 @SpideyMumu both your codes look exactly like mine so I think the issue is somewhere else?

**Comment by eesung00** [comment]
> > Perhaps your Module need to implements keyable too. I did it and it works for module. But I have some issue with the Student class. Anyone knows why? ![image](https://user-images.githubusercontent.com/90175879/159032597-bb81e642-ee35-4e40-90a3-042bb921661a.png) ![image](https://user-images.githubusercontent.com/90175879/159032699-7e69832c-6dda-4d18-b764-98aa3666338c.png)
> 
> I think your error is because your Module should not be implementing Keyable

Oops yeah my bad. After I remove the keyable in module the error still happens but when I  /open Student.java in jshell again it works. Thanks!

**Comment by SpideyMumu** [comment]
I decided to upload on CodeCrunch and the first test case seems to be working. I think it could be some bug in jshell. 

However, now i am getting a different error for student test case.

```
|  Error:

|  incompatible types: KeyableMap<Assessment> cannot be converted to Module
|  new Student("Tony").put(new Module("CS2040").put(new Assessment("Lab1", "B")))
|                          ^---------------------------------------------------^
```

**Comment by LauWyeRock** [comment]
@SpideyMumu Did you make sure your Module::put is correct? I think you're returning KeyableMap<Assessment> instead of Module

**Comment by sherrycqu** [comment]
@SpideyMumu  I think I had the same error as well. Then, I override the put method in Module & Student class by changing the output type to Assessment and Module respectively and it worked.

**Comment by sherrycqu** [comment]
@LauWyeRock  Agree. I think we do not have to 'implement Keyable' again. I guess the reason could be that our Parent class KeyableMap has implemented it so child class should have already inherited it.

**Comment by SpideyMumu** [comment]
> @SpideyMumu I think I had the same error as well. Then, I override the put method in Module & Student class by changing the output type to Assessment and Module respectively and it worked.

Thanks @sherrycqu ! I did make the changes you mentioned and it worked. However, I did break the 'Tell dont ask' principle as I had to make a `getMap()` method in my KeyableMap. Is this the case for you as well?  Below is my Student `put()` method.

```java
 Student put(Module value) {
          return new Student(super.getKey(),
                  super.getMap().put(value.getKey(), value));
      }
```

**Comment by JhChee18** [comment]
> @SpideyMumu I think I had the same error as well. Then, I override the put method in Module & Student class by changing the output type to Assessment and Module respectively and it worked.

I don't think `put(V item)` method in the KeyableMap class is meant to be overridden. Else why codecrunch intentionally mentions that `put(V item)` is one of the two key methods in KeyableMap?

![image](https://user-images.githubusercontent.com/99940180/159117138-29161dce-844b-4597-8214-71efb36cb66e.png)

**Comment by SpideyMumu** [comment]
> > @SpideyMumu I think I had the same error as well. Then, I override the put method in Module & Student class by changing the output type to Assessment and Module respectively and it worked.
> 
> I don't think `put(V item)` method in the KeyableMap class is meant to be overridden. Else why codecrunch intentionally mentions that `put(V item)` is one of the two key methods in KeyableMap?
> 
> ![image](https://user-images.githubusercontent.com/99940180/159117138-29161dce-844b-4597-8214-71efb36cb66e.png)

Makes sense but did you manage to make it work without overriding `put()`? I also think my current way is not the ideal one eventhough it passes all level 2 test cases on CC.

**Comment by Ohnojiso** [comment]
The put method need to be overriden because when you use the KeyableMap put method, it will return a KeyableMap<V> instead of your desired Module/Student/... . I asked how to achieve this in the lab and my TA suggested that I should overload the constructor. You can try having a constructor that takes in a KeyableMap<V>, and in your overriden put method in the subclasses, use that constructor to return the correct type.

**Comment by tungkei** [comment]
I had the same issue for the java.lang.String cannot be converted to java.lang.ClassLoader part and I'm not sure why but the error went away when i reopened the files for ImmutableMap, Assessment and Modules again individually in jshell and ran the test case

**Comment by zeentan22** [comment]
at least for me, I have forgotten to jshell the KeyableMap class so the error appears. Maybe u have forgotten to open the file as well?

**Comment by hongyaong** [comment]
thanks guys the tips n tricks, was wondering if any one knows why, if we override the put method, what is the use for put method in KeyableMap?

**Comment by Swagston20** [comment]
Put is jus a function for KeyableMap. If you dun override it in your subclasses, the output will be KeyableMap instead of Student or Modules.

**Comment by SoSimplySean** [comment]
Great thread - thanks for the help all
